#pragma omp declare target is not allowed in extern "C"

```C++
extern "C" {

#pragma omp declare target
void test()
{
	// does nothing
}
#pragma omp end declare target

}
```

doesn't compile with `clang++ -c -fopenmp target.cpp` and gives

```
target.cpp:3:21: error: directive must be at file or namespace scope
#pragma omp declare target
                    ^
target.cpp:8:25: error: unexpected OpenMP directive '#pragma omp end declare target'
#pragma omp end declare target
                        ^
2 errors generated.
```

--- a/lib/Sema/SemaOpenMP.cpp
+++ b/lib/Sema/SemaOpenMP.cpp
@@ -1564,7 +1564,8 @@ void Sema::CompleteOMPDeclareReductionDecl(OMPDeclareReductionDecl *D,
 
 bool Sema::ActOnStartOpenMPDeclareTargetDirective(Scope *S,
                                                   SourceLocation Loc) {
-  if (CurContext && !CurContext->isFileContext()) {
+  if (CurContext && !CurContext->isFileContext() &&
+      !CurContext->isExternCContext() && !CurContext->isExternCXXContext()) {
     Diag(Loc, diag::err_omp_region_not_file_context);
     return false;
   }

--- a/test/OpenMP/declare_target_ast_print.cpp
+++ b/test/OpenMP/declare_target_ast_print.cpp
@@ -15,8 +15,32 @@ void foo() {}
 #pragma omp end declare target
 // CHECK: #pragma omp end declare target
 
+extern "C" {
+#pragma omp declare target
+// CHECK: #pragma omp declare target
+
+void foo_c() {}
+// CHECK-NEXT: void foo_c()
+
+#pragma omp end declare target
+// CHECK: #pragma omp end declare target
+}
+
+extern "C++" {
+#pragma omp declare target
+// CHECK: #pragma omp declare target
+
+void foo_cpp() {}
+// CHECK-NEXT: void foo_cpp()
+
+#pragma omp end declare target
+// CHECK: #pragma omp end declare target
+}
+
 int main (int argc, char **argv) {
   foo();
+  foo_c();
+  foo_cpp();
   return (0);
 }
 

